mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -128,22 +148,166 @@ void affixThreadwiseCopyV2Attributes(ThreadwiseCopyV2Op top,
+            scatterLoop.getLowerCoords(1)[0]);
+        b.create<miopen::YieldOp>(loc, toYieldInner);
+      }
+      toYield = scatterLoop.getResults()[0];
+    } else {
+      toYield = b.create<InsertSliceOp>(loc, resultType, loaded, loopArg,
+                                        loop.getLowerCoords(1)[0]);
+    }
+  }
+  b.create<miopen::YieldOp>(loc, toYield);
+  return loop;
 }
 
-template <typename T, typename U>
-void affixBlockwiseCopyAttributes(T &bop, U &gop, OpBuilder &b, int vectorDim,
-                                  int blockwiseLoadLength,
-                                  int blockwiseStoreLength) {
-  bop->setAttr("block_size", gop->getAttr("block_size"));
-
-  bop->setAttr("source_vector_read_dim", b.getI32IntegerAttr(vectorDim));
-  bop->setAttr("dest_vector_write_dim", b.getI32IntegerAttr(vectorDim));
-  bop->setAttr("source_data_per_read",
-               b.getI32IntegerAttr(blockwiseLoadLength));
-  bop->setAttr("dest_data_per_write",
-               b.getI32IntegerAttr(blockwiseStoreLength));
+TransformingForOp createLdsStoreLoop(OpBuilder &b, Location loc, Value loaded,
+                                     Value buffer, ValueRange bufferStart,
+                                     Type storingType,
+                                     ArrayRef<int64_t> sliceLengths,
+                                     uint32_t vectorDim) {
+  Type loadedType = loaded.getType();
+  bool fullyScalar = !loadedType.isa<ShapedType>();
+
+  int64_t storeLength = 1;
+  if (auto storingVectorType = storingType.dyn_cast<VectorType>())
+    storeLength = storingVectorType.getNumElements();
+
+  size_t nUpper = bufferStart.size();
+  bool complexVectorStore = (storeLength > 1) && (vectorDim != nUpper - 1);
+
+  Value zero = b.createOrFold<ConstantIndexOp>(loc, 0);
+  SmallVector<Value, 5> linearInit(nUpper, zero);
+
+  ArrayAttr bufferTransforms;
+  buffer = untransform(b, buffer, bufferTransforms);
+  ArrayAttr noTransforms = b.getArrayAttr({});
+  ArrayAttr resultIdxMap = makeLinearDomain(b, loc, sliceLengths);
+
+  SmallVector<int64_t, 4> loopBounds;
+  llvm::copy(sliceLengths, std::back_inserter(loopBounds));
+  assert(loopBounds[vectorDim] % storeLength == 0 && "Uneven vector store");
+  loopBounds[vectorDim] /= storeLength;
+
+  SmallVector<Attribute> loopTransforms = {resultIdxMap, bufferTransforms};
+  if (complexVectorStore)
+    loopTransforms[0] = noTransforms;
+
+  auto loop = b.create<TransformingForOp>(
+      loc, ArrayRef<ValueRange>{linearInit, bufferStart}, loopTransforms,
+      loopBounds,
+      /*forceUnroll=*/true, /*useIndexDiffs=*/true);
+  OpBuilder::InsertionGuard guard(b);
+  b.setInsertionPointToStart(loop.getBody());
+
+  // We can use vector.transfer_write if the vectorized dimension of the
+  // load-store loops is the fastest-moving dimension of the store loop.
+  // Otherwise, we need to gather elements from the result vector in order to
+  // store them
+  if (fullyScalar)
```

Comment:
This is a 3-pronged if statement, let's use curly braces here to conform with other branches.